### ISSUE TITLE
[BUG] `DtwPythonDist` output has randomly permuted rows/columns

### ISSUE THREAD

**Issue by stephanielees** [sktime/sktime#5988]
Hi,

I tried using DtwPythonDist() to calculate distances between a dataset of time series. It worked well, but I noticed that I get different results every time I run it. 

Here's my code. I run it in my local machine using PyCharm.

`dtwpy = DtwPythonDist()`
`dist_mat = dtwpy.transform(data)`

I used it with two different datasets: the dataset with all equal lengths, and the dataset with unequal lengths. The DtwPythonDist() behaves the same. If I run it twice, I would get two totally different distance matrices.

Am I missing something? What should I do?

Thank you.

**Comment by fkiraly** [comment]
Hm, that's odd, I thought the algorithm was deterministic.

Could you provide a more complete example, e.g., with dummy data, that has the same property - run twice it will produce different distances?

**Comment by stephanielees** [comment]
I just tried it with a dummy dataset with only three sequences. Here is the code:

`import dtw`
`import sktime`
`from sktime.dists_kernels.dtw import DtwPythonDist`

`rng = np.random.default_rng(1)`
`cities = np.append(np.repeat('a', 9), np.repeat('b', 12))`
`cities = np.append(cities, np.repeat('c', 12))`
`dates = pd.date_range(start='1-1-2020', end='10-1-2020', freq='M').append(pd.date_range('1-1-2020', '1-1-2021', freq='M'))`
`dates = dates.append(pd.date_range('1-1-2020', '1-1-2021', freq='M'))`

`myid = pd.MultiIndex.from_tuples(list(zip(cities, dates)))`
`df = pd.DataFrame(data={'val': rng.uniform(size=33)}, index=myid)`

`dtwpy = DtwPythonDist()`
`dtwpy.transform(df)`

When I first run the code above, I got

> [[0.        , 2.65004765, 3.29971253],
       [2.65004765, 0.        , 3.19721144],
       [3.29971253, 3.19721144, 0.        ]]

But when I run it again, I got

> [[0.        , 3.19721144, 2.65004765],
       [3.19721144, 0.        , 3.29971253],
       [2.65004765, 3.29971253, 0.        ]]

The numbers seem to be permutated somehow. In my real project, I have 19 sequences, and the permutation is a bit more random. 

If I understand the result correctly, the entry (i, j) of a distance matrix represents the distance between data[i] and data[j], right? So in the first matrix, it shows that the distance between the first series and the third series in the dataset is 3.2997, while in the second matrix, the distance between them is 2.65.

I only tried with only three sequences for this example, but feel free to let me know if you need more.

**Comment by fkiraly** [comment]
Thanks, that is useful!

FYI, I've formatted the code a bit, you can check dots (top right) -> edit to see how to paste full code blocks.

It seems odd that the result is non-deterministic. From my understanding of the dynamic programming algorithm, the dtw distance should be fully deterministic, even if the underlying path may be ambiguous.

So, this sounds like a bug.

**Comment by fkiraly** [comment]
FYI, `sktime` has two more implementations of the DTW distance available:

* `DtwDistTslearn`, from `tslearn`
* `DtwDist`, an implementation directly in `sktime`

**Comment by stephanielees** [comment]
I tried computing distance between `df.loc['a']` and `df.loc['b']` using `dtw` from `dtw-python` directly, and it gives the same result when I run it twice. But not with the `DtwPythonDist`.

Regarding other implementations from `sktime`, I haven't tried out `DtwDistTslearn` yet because I want to implement Derivative DTW. So far, I've only tried `DtwPytonDist` and `DtwDist`. `DtwDist` doesn't behave strangely, but I want to compute distances between series with unequal lengths. So, it is unfortunately out of option.

**Comment by fkiraly** [comment]
thanks. What would help if you would paste your code when you directly try `dtw-python`, and the distances you get - let's check if the parameters are the same.

Regarding derivative dtw, is that not the same as first differences, then dtw? If so, you can use `Differencer` and chain it with any `dtw` to get derivative dtw:

```python
from sktime.transformations.series.difference import Differencer

my_ddtw = Differencer() * my_dtw
```
where `my_dtw` is any dtw algorithm.

**Comment by fkiraly** [comment]
@stephanielees, I cannot reproduce the error, I have executed your code 10 times, and each time I get the result

```
array([[0.        , 2.65004765, 3.19721144],
       [2.65004765, 0.        , 3.29971253],
       [3.19721144, 3.29971253, 0.        ]])
```

I am using python 3.11, windows, `dtw-python 3.11`, and `sktime 0.26.0`.

**Comment by fkiraly** [comment]
> The numbers seem to be permutated somehow.

Looking at your matrices, this would be consistent with the individual series getting permuted. From your first matrix, sequences 0, 1, 2 change order to 1, 2, 0, i.e., the first series moves to the end of the ordering.

**Comment by stephanielees** [comment]
> Can you please follow the instructions on reporting your versions (from the "report bug" template) - run `show_versions`. I'll also need your `dtw-python` version.

Oops, I just realized I haven't given that information. Sorry about that. My sktime `sktime` version is 0.26.0, `dtw-python` is 1.3.1, python 3.11, windows (so it's the same environment as yours). I run this code in Pycharm.
 
> Also, just to check, you are using `dtw-python`, not `dtw`, right?
> Unfortunately these packages may be easy to confuse, due to the error message of a failed import, `dtw` - but the interfaced package is `dtw-python`, which is also imported by `import dtw`.

Yes, I installed the `dtw-python` library, then imported it as `dtw`.

**Comment by stephanielees** [comment]
> > The numbers seem to be permutated somehow.
> 
> Looking at your matrices, this would be consistent with the individual series getting permuted. From your first matrix, sequences 0, 1, 2 change order to 1, 2, 0, i.e., the first series moves to the end of the ordering.

So, is it problematic? Because the ultimate goal is to use this distance matrix for clustering. When I use a dummy, fixed distance matrix, I always get the same clustering result. But when I calculate the distance with `DtwPythonDist`, I could get very different clusters every time.

**Comment by fkiraly** [comment]
> So, is it problematic?

Well, the set of clusters won't change, you can just not link them to the original series anymore.

Let's try to diagnose this:

Can you please paste your `show_versions` output as described in the bug report instructions? (you can open a new bug report to read these instructions)

I am starting to suspect it might have to do with `pandas` or similar.

**Comment by fkiraly** [comment]
Can you also try to execute multiple times:
`df.get_level_values(0)`, and: (a) print what you get; (b) check whether that changes every time you run it?

**Comment by fkiraly** [comment]
For the last snippet, I always get `['c', 'b', 'a']`, here's my `show_versions`:

```
System:
    python: 3.11.5 | packaged by Anaconda, Inc. | (main, Sep 11 2023, 13:26:23) [MSC v.1916 64 bit (AMD64)]
   machine: Windows-10-10.0.22621-SP0

Python dependencies:
          pip: 23.3.1
       sktime: 0.26.0
      sklearn: 1.3.2
       skbase: 0.7.2
        numpy: 1.26.2
        scipy: 1.11.4
       pandas: 2.2.0
   matplotlib: 3.8.3
       joblib: 1.3.2
        numba: None
  statsmodels: 0.14.0
     pmdarima: 2.0.4
statsforecast: None
      tsfresh: None
      tslearn: None
        torch: None
   tensorflow: None
tensorflow_probability: None
```

**Comment by stephanielees** [comment]
Hi @fkiraly , I'm sorry for my late reply.

> Can you also try to execute multiple times (in your example): `df.get_level_values(0)`, and: (a) print what you get; (b) check whether that changes every time you run it?
> 
> & if that is fine, then this code:
> 
> ```python
> ix = set(df.index.get_level_values(0))
> [i for i in ix]
> ```
> 
> again, (a) print what you get; (b) check whether it changes each time you run

I always get the same result for `df.get_level_values(0)`. For `ix = set(df.index.get_level_values(0))`, I sometimes get a different result.

Here's my `show_versions`:
```
System:
    python: 3.11.4 (tags/v3.11.4:d2340ef, Jun  7 2023, 05:45:37) [MSC v.1934 64 bit (AMD64)]
   machine: Windows-10-10.0.19045-SP0

Python dependencies:
          pip: 22.3.1
       sktime: 0.26.0
      sklearn: 1.4.0
       skbase: 0.7.2
        numpy: 1.26.4
        scipy: 1.12.0
       pandas: 2.1.4
   matplotlib: 3.8.3
       joblib: 1.3.2
        numba: 0.59.0
  statsmodels: None
     pmdarima: None
statsforecast: None
      tsfresh: None
      tslearn: None
        torch: None
  tensorflow: None
tensorflow_probability: None
```

**Comment by fkiraly** [comment]
> For `ix = set(df.index.get_level_values(0))`, I sometimes get a different result.

I see - I can't reproduce it, but in that case this PR should fix your problem:
https://github.com/sktime/sktime/pull/6007

**Comment by fkiraly** [comment]
The fix should be in version 0.27.1.

It would be appreciated, if until then you could check whether it fixes the problem for you (e.g., by installing the development version https://www.sktime.net/en/stable/installation.html#development-versions), because we currently do no thave a means to reproduce the issue.

**Comment by stephanielees** [comment]
Hi, I'm really sorry for the late reply. I just had the time to try it again with sktime version 0.29.0. It works now; the results don't permute anymore. Thank you for your help.
